### PR TITLE
[Prelude 2]: VenusOrbitalSurvey

### PR DESCRIPTION
**Pull request by eratchia**
Not sure whether the chosen bought cards should be known (if one of two non-venus cards is bought). For now I left it unknown because it was easier this way.

### CODE REVIEW

**Comment by kberg** [comment]
Thank you! I'll look at this shortly.

File: src/server/cards/prelude2/VenusOrbitalSurvey.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        renderData: CardRenderer.builder((b) => {
+          b.action(undefined, (ab) =>
+            ab.empty().startAction.empty()).text('Action: Reveal the top 2 cards, take any venus cards to hand for free. Any other card you either buy or discard', Size.TINY, true);
+        }),
+      },
+    });
+  }
+
+  public canAct(player: IPlayer): boolean {
+    if (!player.game.projectDeck.canDraw(2)) {
```
Comment (kberg) [inline src/server/cards/prelude2/VenusOrbitalSurvey.ts:34]: I can't recall if there's been discussion about what should happen in this case (or even if some solutions have been implemented) but I believe the answer is, if the cards can't be drawn, the action can't be taken.

Save me from doing research, I suppose? :D

File: src/server/cards/prelude2/VenusOrbitalSurvey.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+      this.warnings.add('deckTooSmall');
+    }
+    return true;
+  }
+
+  public action(player: IPlayer) {
+    const isVenus = (card: IProjectCard) => player.tags.cardHasTag(card, Tag.VENUS);
+
+    player.game.defer(new DrawCards(player, 2, {}).andThen((cards) => {
```
Comment (kberg) [inline src/server/cards/prelude2/VenusOrbitalSurvey.ts:43]: , {} is not required.

File: src/server/cards/prelude2/VenusOrbitalSurvey.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      let message = '${0} revealed ';
+      if (cards.length === 0) {
```
Comment (kberg) [inline src/server/cards/prelude2/VenusOrbitalSurvey.ts:45]: Yes, cards should be revealed.

No need for custom logging, use LogHelper.logDrawnCards (and don't set privateMessage)

**Review by kberg — COMMENTED** [review]
This is nice work.

Cards being public should simplify the second half of the function too.

File: src/server/cards/prelude2/VenusOrbitalSurvey.ts (hunk 3 above)
Comment (eratchia) [reply src/server/cards/prelude2/VenusOrbitalSurvey.ts:45]: I think the distinction is important here as not all of the revealed cards land in the players hand here.

File: src/server/cards/prelude2/VenusOrbitalSurvey.ts (hunk 1 above)
Comment (eratchia) [reply src/server/cards/prelude2/VenusOrbitalSurvey.ts:34]: I was looking at it like something between drawing cards and things like search for life/asteroid deflection system that reveal cards. Both seem to be available with lacking cards with the latter giving a warning.

**Comment by kberg** [comment]
Thanks for this. At this point, I'm trying to just simplify your implementation. I'm not going to force  those changes, but I do think we'll wind up down the road removing that behavior from SearchForLife, for instance.

**Comment by eratchia** [comment]
Well in that case, I think allowing for the action anyways without any warning is the way to go.

**Comment by kberg** [comment]
> Well in that case, I think allowing for the action anyways without any warning is the way to go.

OK. What I mean though is eventually Search For Life's action won't be playable if the draw and discard piles are empty. (e.g. you can't use it as a stall action.)

**Comment by eratchia** [comment]
> > Well in that case, I think allowing for the action anyways without any warning is the way to go.
> 
> OK. What I mean though is eventually Search For Life's action won't be playable if the draw and discard piles are empty. (e.g. you can't use it as a stall action.)

Oh, ok. But I assume this action should be still usable with one card right?

**Comment by kberg** [comment]
> Oh, ok. But I assume this action should be still usable with one card right?

I don't think so. Here's why:

It is generally known (though I didn't really internalize it) that there are only three things that are optional in an action:
1. If a global parameter is at its maximum, you can take an action that raises a global parameter without having to do so. So, spending 8 heat when temperature is already maxxed is legal, but used as a stall tactic.
2. Actions that remove resources from any player is optional: you do not have to remove resources. Side-note: removing production _is_ mandatatory. So if an action removes plant production and you're the only one with plant production, guess whose plant production gets removed.
3. Actions that add resources of a certain type to your cards are allowed, even if you don't have any cards of that type.

ref: https://boardgamegeek.com/thread/1705209/official-faq-rules-clarifications

I tried to get a statement about running out the deck, but my question was not taken very seriously: https://boardgamegeek.com/thread/3217206/empty-deck-discard-pile-during-action-phase/page/1

You can see that my original thought was to house rule to allow the action, but that's just because the app hasn't taken a definitive stance yet.

So like I say, leave it be if you like. But if you want to keep the code simpler, take it out.

Either way, at least you understand where I'm coming from.

**Comment by eratchia** [comment]
Huh, thanks for the detailed explanation. I didn't previously understand drawing to be manatory.

I do think card draw being required does make low deck gameplay very weird but I can definitely understand that the readability and rule consistency shouldn't suffer because of such an edge case.

File: src/server/cards/prelude2/VenusOrbitalSurvey.ts [hunk 4]
```diff
@@ -31,31 +31,17 @@ export class VenusOrbitalSurvey extends Card implements IActionCard {
...
-        }
-      }
+      const message = '${0} revealed ${1} and ${2}';
       player.game.log(message, (b) => {
-        b.player(player);
-        for (const card of cards) {
-          b.card(card);
-        }
-      }, {});
+        b.player(player).card(cards[0]).card(cards[1]);
```
Comment (kberg) [inline src/server/cards/prelude2/VenusOrbitalSurvey.ts:43]: And add the {tags: true} attribute for both cards, just like SearchForLife. It'll be useful.

`player.game.log('${0} revealed and discarded ${1}', (b) => b.player(player).card(card, {tags: true}));`

**Review by kberg — APPROVED** [review]
Add that last bit if you don't mind. This will be good to deploy tomorrow. Thank you!

**Comment by kberg** [comment]
Thanks for sticking with me.

**Pull request merged**